File: src/scaling.rs
use std::time::Instant;
use std::fs;
use std::path::Path;
use csv::Writer;
use rayon::ThreadPoolBuilder;
use crate::grid::Grid;
use crate::{par_sim, seq_sim};


pub struct ScalingResult {
    pub experiment: String,
    pub threads: usize,
    pub width: usize,
    pub height: usize,
    pub steps: usize,
    pub run: usize,
    pub duration_ms: f64
}

fn time_seq(width: usize, height: usize, steps: usize) -> f64 {
    let mut grid = Grid::new_with_params(width, height, 0.7, 0.7);
    let start = Instant::now();
    seq_sim::step_n(&mut grid, steps);
    start.elapsed().as_secs_f64() * 1000.0
}


fn time_par(width: usize, height: usize, steps: usize, threads: usize) -> f64 {
    // build a rayon thread pool with exactly `threads` threads
    let pool = ThreadPoolBuilder::new()
        .num_threads(threads)
        .build()
        .expect("Failed to build thread pool");

    let mut grid = Grid::new_with_params(width, height, 0.7, 0.7);
    let start = Instant::now();
    pool.install(|| par_sim::step_n(&mut grid, steps));
    start.elapsed().as_secs_f64() * 1000.0
}

pub fn run_strong_scaling(
    width: usize,
    height: usize,
    steps: usize,
    max_threads: usize,
    runs: usize,
    output_path: &str,
) {
    fs::create_dir_all(Path::new(output_path).parent().unwrap_or(Path::new(".")))
        .expect("Failed to create output directory");

    let mut writer = Writer::from_path(output_path)
        .expect("Failed to create CSV");

    writer.write_record(&[
        "experiment", "threads", "width", "height", "steps", "run", "duration_ms"
    ]).unwrap();

    println!("\nStrong scaling — fixed grid {}x{}, {} steps", width, height, steps);
    println!("(same problem size, more threads)\n");

    // sequential baseline — recorded as threads=1
    for run in 0..runs {
        let ms = time_seq(width, height, steps);
        writer.write_record(&[
            "strong_sequential",
            "1",
            &width.to_string(),
            &height.to_string(),
            &steps.to_string(),
            &run.to_string(),
            &format!("{:.3}", ms),
        ]).unwrap();
    }
    println!("  sequential baseline: done ({} runs)", runs);

    // parallel with 2..=max_threads
    for threads in 2..=max_threads {
        for run in 0..runs {
            let ms = time_par(width, height, steps, threads);
            writer.write_record(&[
                "strong_parallel",
                &threads.to_string(),
                &width.to_string(),
                &height.to_string(),
                &steps.to_string(),
                &run.to_string(),
                &format!("{:.3}", ms),
            ]).unwrap();
        }
        println!("  {} threads: done ({} runs)", threads, runs);
        writer.flush().unwrap();
    }

    println!("Strong scaling results saved to: {}", output_path);
}

pub fn run_weak_scaling(
    base_width: usize,
    base_height: usize,
    steps: usize,
    max_threads: usize,
    runs: usize,
    output_path: &str,
) {
    fs::create_dir_all(Path::new(output_path).parent().unwrap_or(Path::new(".")))
        .expect("Failed to create output directory");

    let mut writer = Writer::from_path(output_path)
        .expect("Failed to create CSV");

    writer.write_record(&[
        "experiment", "threads", "width", "height", "steps", "run", "duration_ms"
    ]).unwrap();

    println!("\nWeak scaling — base grid {}x{}, {} steps", base_width, base_height, steps);
    println!("(problem grows with thread count, constant work per thread)\n");

    // sequential baseline at base size
    for run in 0..runs {
        let ms = time_seq(base_width, base_height, steps);
        writer.write_record(&[
            "weak_sequential",
            "1",
            &base_width.to_string(),
            &base_height.to_string(),
            &steps.to_string(),
            &run.to_string(),
            &format!("{:.3}", ms),
        ]).unwrap();
    }
    println!("  1 thread ({}x{}): done", base_width, base_height);

    // for weak scaling: multiply height by thread count to keep work-per-thread constant
    for threads in 2..=max_threads {
        let width  = base_width;
        let height = base_height * threads; // scale problem with threads

        for run in 0..runs {
            let ms = time_par(width, height, steps, threads);
            writer.write_record(&[
                "weak_parallel",
                &threads.to_string(),
                &width.to_string(),
                &height.to_string(),
                &steps.to_string(),
                &run.to_string(),
                &format!("{:.3}", ms),
            ]).unwrap();
        }
        println!("  {} threads ({}x{}): done", threads, width, height);
        writer.flush().unwrap();
    }

    println!("Weak scaling results saved to: {}", output_path);
}

File: python/scaling.py
import time
import csv
import os
import multiprocessing as mp
import grid
import seq_sim
import par_sim

def run_strong_scaling(width, height, steps, max_workers, runs, output_path):
    os.makedirs(os.path.dirname(output_path), exist_ok=True)

    print(f'\nStrong scaling — fixed grid {width}x{height}, {steps} steps')
    print('(same problem size, more workers)\n')

    with open(output_path, 'w', newline='') as f:
        writer = csv.writer(f)
        writer.writerow(['experiment', 'workers', 'width', 'height', 'steps', 'run', 'duration_ms'])

        #sequential baseline
        for run in range(runs):
            rust, hum, oxy = grid.new_grid(width, height, 0.7, 0.7)
            start = time.perf_counter()
            seq_sim.step_n(rust, hum, oxy, steps)
            ms = (time.perf_counter() - start) * 1000
            writer.writerow(['strong_sequential', 1, width, height, steps, run, f'{ms:.3f}'])
        print(f'  sequential baseline: done ({runs} runs)')

        #parallel with increasing workers
        for workers in range(2, max_workers + 1):
            for run in range(runs):
                rust, hum, oxy = grid.new_grid(width, height, 0.7, 0.7)
                start = time.perf_counter()
                par_sim.step_n(rust, hum, oxy, steps, workers)
                ms = (time.perf_counter() - start) * 1000
                writer.writerow(['strong_parallel', workers, width, height, steps, run, f'{ms:.3f}'])
            print(f'  {workers} workers: done ({runs} runs)')
            f.flush()

    print(f'Strong scaling results saved to: {output_path}')

def run_weak_scaling(base_width, base_height, steps, max_workers, runs, output_path):
    os.makedirs(os.path.dirname(output_path), exist_ok=True)

    print(f'\nWeak scaling — base grid {base_width}x{base_height}, {steps} steps')
    print('(problem grows with workers, constant work per worker)\n')

    with open(output_path, 'w', newline='') as f:
        writer = csv.writer(f)
        writer.writerow(['experiment', 'workers', 'width', 'height', 'steps', 'run', 'duration_ms'])

        #sequential baseline at base size
        for run in range(runs):
            rust, hum, oxy = grid.new_grid(base_width, base_height, 0.7, 0.7)
            start = time.perf_counter()
            seq_sim.step_n(rust, hum, oxy, steps)
            ms = (time.perf_counter() - start) * 1000
            writer.writerow(['weak_sequential', 1, base_width, base_height, steps, run, f'{ms:.3f}'])
        print(f'  1 worker ({base_width}x{base_height}): done')

        for workers in range(2, max_workers + 1):
            width  = base_width
            height = base_height * workers
            for run in range(runs):
                rust, hum, oxy = grid.new_grid(width, height, 0.7, 0.7)
                start = time.perf_counter()
                par_sim.step_n(rust, hum, oxy, steps, workers)
                ms = (time.perf_counter() - start) * 1000
                writer.writerow(['weak_parallel', workers, width, height, steps, run, f'{ms:.3f}'])
            print(f'  {workers} workers ({width}x{height}): done')
            f.flush()

    print(f'Weak scaling results saved to: {output_path}')